crawler/index.ts

```diff
@@ -189,86 +209,102 @@ program
         if (isNaN(batchId)) {
           logger.error('Invalid batch ID %s', options.batch_id);
           throw new Error('Invalid batch ID');
         }
-        const row = db
-          .query<
-            { height: number; ts: number },
-            [number, string]
-          >('SELECT height, ts FROM tasks WHERE batch_id = ? AND protocol_id = ? AND status = "new" ORDER BY batch_id ASC LIMIT 1')
-          .get(batchId, protocolId);
-        if (!row) {
+        const { rows } = await db.query(
+          'SELECT height, ts FROM tasks WHERE batch_id = $1 AND protocol_id = $2 AND status = $3 ORDER BY batch_id ASC LIMIT 1',
+          [batchId, protocolId, 'new'],
+        );
+        if (rows.length === 0) {
           logger.info('No tasks found for batch_id %s', options.batch_id);
           throw new Error('No tasks found');
         }
-        return { batchId, height: row.height };
+        return { batchId, height: rows[0].height };
       } else {
-        const row = db
-          .query<
-            { height: number; batch_id: number; ts: number },
-            string
-          >('SELECT height, batch_id, ts FROM tasks WHERE protocol_id = ? AND status = "new" ORDER BY batch_id ASC LIMIT 1')
-          .get(protocolId);
-        if (!row) {
+        const { rows } = await db.query(
+          'SELECT height, batch_id, ts FROM tasks WHERE protocol_id = $1 AND status = $2 ORDER BY batch_id ASC LIMIT 1',
+          [protocolId, 'new'],
+        );
+        if (rows.length === 0) {
           logger.info('No tasks found for protocol %s', protocolId);
           throw new Error('No tasks found');
         }
         return {
-          batchId: row.batch_id,
-          height: row.height,
+          batchId: rows[0].batch_id,
+          height: rows[0].height,
         };
       }
     })();
-    const multipliers = getAssetMulsByProtocolAndBatchId(protocolId, batchId);
+
+    const multipliers = await getAssetMulsByProtocolAndBatchId(
+      db,
+      protocolId,
+      batchId,
+    );
     logger.info(
       'Processing task for protocol %s, height %d and batch_id %d multipliers %o',
       protocolId,
       height,
       batchId,
       multipliers,
     );
+
     // Processing the source
     const sourceObj = new sources[
       config.protocols[protocolId].source as keyof typeof sources
     ](config.protocols[protocolId].rpc, logger, config.protocols[protocolId]);
+
     await sourceObj.getUsersBalances(
       height,
       multipliers,
-      (balances: UserBalance[]) => {
-        const query = db.prepare<
-          unknown,
-          [number, string, string, number, string, string]
-        >(
-          'INSERT INTO user_data (batch_id, address, protocol_id, height, asset, balance) VALUES (?, ?, ?, ?, ?, ?);',
-        );
-        const insert = db.transaction((balances) => {
+      async (balances: UserBalance[]) => {
+        const callbackDbClient = await connect(false, config, logger);
```

Comment:
this can lead to pool exhaustion, please make sure connection pool size is like around 20-30 or so

Reply:
Using a connection pool with 30 connections now